logical/plugin/plugin.go

```diff
@@ -0,0 +1,75 @@
+package plugin
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/vault/helper/pluginutil"
+	"github.com/hashicorp/vault/logical"
+)
+
+// PluginMap should be used by clients for map of plugins
+var PluginMap = map[string]plugin.Plugin{
+	"backend": &BackendPlugin{},
+}
+
+// NewBackend will return an instance of an RPC-based client implementation of the backend for
+// external plugins, or a concrete implementation of the backend if it is a builtin backend.
+// The backend is returned as a logical.Backend interface.
+func NewBackend(pluginName string, sys pluginutil.LookRunnerUtil) (logical.Backend, error) {
+	// Look for plugin in the plugin catalog
+	pluginRunner, err := sys.LookupPlugin(pluginName)
+	if err != nil {
+		return nil, err
+	}
+
+	var backend logical.Backend
+	if pluginRunner.Builtin {
+		// Plugin is builtin so we can retrieve an instance of the interface
+		// from the pluginRunner. Then cast it to a Database.
+		backendRaw, err := pluginRunner.BuiltinFactory()
+		if err != nil {
+			return nil, fmt.Errorf("error getting plugin type: %s", err)
+		}
+
+		var ok bool
+		backend, ok = backendRaw.(logical.Backend)
+		if !ok {
+			return nil, fmt.Errorf("unsuported backend type: %s", pluginName)
+		}
+
+	} else {
+		// create a DatabasePluginClient instance
+		backend, err = newPluginClient(sys, pluginRunner)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return backend, nil
+}
+
+func newPluginClient(sys pluginutil.RunnerUtil, pluginRunner *pluginutil.PluginRunner) (logical.Backend, error) {
+	client, err := pluginRunner.Run(sys, PluginMap, handshakeConfig, []string{})
```

Comment:
I added a wrapper around `backendPluginClient`, similar to the example you provided. It should be noted that, however, "proper" cleanup (i.e. `client.Kill()`) is only done when a client is instantiated via `newPluginClient()` since `backendPluginClient` has no context of the `plugin.Client` that spawned it.

This means that [backend_test.go](https://github.com/hashicorp/vault/pull/2874/files#diff-14589062022109543eda6861585020a7R151) won't be able to test cleanup for `plugin.Client.Kill()` since it uses `TestPluginGRPCConn` to get an RPC client and server directly.